game-core/src/test/java/games/strategy/triplea/attachments/TriggerAttachmentTest.java

```diff
@@ -26,102 +29,226 @@
-    final TriggerAttachment triggerAttachment = mock(TriggerAttachment.class);
-    final Set<TriggerAttachment> satisfiedTriggers = Collections.singleton(triggerAttachment);
-
-    when(triggerAttachment.getPropertyMap()).thenCallRealMethod();
-    triggerAttachment.getPropertyMap().get("playerAttachmentName").setValue("rulesAttachment:RulesAttachment");
-    when(triggerAttachment.getPlayerProperty())
-        .thenReturn(Collections.singletonList(
-            Tuple.of("productionPerXTerritories", "someNewValue")));
-    when(triggerAttachment.getName()).thenReturn("mockedTriggerAttachment");
-
-    final PlayerId playerId = mock(PlayerId.class);
-    when(playerId.getAttachment("rulesAttachment"))
-        .thenReturn(new RulesAttachment(null, null, gameData));
-    when(triggerAttachment.getPlayers()).thenReturn(Collections.singletonList(playerId));
-
-
-    TriggerAttachment.triggerPlayerPropertyChange(
-        satisfiedTriggers,
-        bridge,
-        "beforeOrAfter",
-        "stepName",
-        false, // useUses
-        false, // testUses
-        false, // testChance
-        false); // testWhen
-    verify(bridge).addChange(not(argThat(Change::isEmpty)));
+  void testGetPropertyChangeHistoryStartEvent() {
+    final BiFunction<String, String, Optional<Tuple<Change, String>>> fun =
+        (startValue, newValue) -> {
+
+          final TriggerAttachment triggerAttachment = new TriggerAttachment("aTriggerAName", null, null);
+          final TestAttachment propertyAttachment = new TestAttachment("aTestAName", null, null);
+          propertyAttachment.setValue(startValue);
+          final Named attachedTo = mock(Named.class);
+
+          return TriggerAttachment.getPropertyChangeHistoryStartEvent(
+              triggerAttachment, propertyAttachment,
+              "value", // Property name in 'TestAttachment'. Authentic name: "productionPerXTerritories".
+              Tuple.of(true, newValue),
+              "rulesAttachment", attachedTo);
+        };
+
+    {
+      final Optional<Tuple<Change, String>> r = fun.apply("2:conscript", "4:conscript");
+
+      assertTrue(r.isPresent());
+      assertFalse(r.get().getFirst().isEmpty());
+      assertFalse(r.get().getSecond().isEmpty());
+    }
+
+    {
+      final Optional<Tuple<Change, String>> r = fun.apply("", "4:conscript");
+
+      assertTrue(r.isPresent());
+      assertFalse(r.get().getFirst().isEmpty());
+      assertFalse(r.get().getSecond().isEmpty());
+    }
+
+    {
+      final Optional<Tuple<Change, String>> r = fun.apply("4:conscript", "4:conscript");
+
+      assertFalse(r.isPresent());
+    }
+
+    {
+      final Optional<Tuple<Change, String>> r = fun.apply("4:conscript", "");
+
+      assertTrue(r.isPresent());
+      assertFalse(r.get().getFirst().isEmpty());
+      assertFalse(r.get().getSecond().isEmpty());
+    }
+
+    {
+      final Optional<Tuple<Change, String>> r = fun.apply("", "");
+
+      assertFalse(r.isPresent());
+    }
   }
```

Comment:
For this Test case, I'd recommend creating a nested test class using `@Nested` where each of those blocks are individual test cases and the BiFunction is actually a real helper method.